### PR TITLE
add combat ai based on threattables for both raiders and guards

### LINKED ISSUE #6841: Guards take too slow to attack a target
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version:
- MineColonies Version:
- Structurize Version:
- Related Mods and their Versions:

## Expected behavior
Guards will respond better to moving target that is away from them

## Actual behavior
Sometimes we will see citizens getting killed and the guard being a braindead and ignore the help, or being wonky in general.

## Steps to reproduce the issue

Here's an incident I had that ticks me to make this thread.
My citizen got attacked by baby zombie and called for help, ran away for quite a distance, then a knight arrive as the two run toward him, but to my surprise, the knight completely ignored the chase to travel to the help destination, and by the time he did the chase got escalated to the opposite of town, and nothing got solved.

### Logs

- latest.log: 
- crashlog: 

### Notes

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7450: A few issues with guards in following mode
### Prerequisites 

- [X] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [X] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [X] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version:
- MineColonies Version:
- Structurize Version:
- Related Mods and their Versions:

## Expected behavior

+ Guards would immediately go after any mobs that hit you
+ Guards would pander around a distance around you
+ Guards would immediately teleport after you

## Actual behavior

- Guards don't attack mobs that hit you, you would have to hit the mob first.
- Guards got carried away too much with auto-targeting mobs and eventually stray too far from you, even in tight grouping
- Guards take awhile to decide to teleport to you from distances away, it is maybe due to their excessive auto-targeting mechanic
- If the mob is not on the list of the guards, or isn't a hostile mobs(i.e:I/F dragons.) then guards will completely ignore when they hit you or you hit them.

## Steps to reproduce the issue

1. Select tight grouping following mode
2. Lead your guard(s) into the mob battlefield at night 
3. Observe the guard mechanic with the issues listed
One time I was spited by a Hydra from a distance and the guards were just ignoring it, if I remember correctly then even after I punched the Hydra, guards still don't attack it despite right next to it, just because at that time Hydra wasn't exist on their mob list

### Logs

- latest.log: 
- crashlog: 

### Notes

Basically I want the guards when following to be as loyal as Minecraft wolves, when following they should switch to protecting their leader instead of citizens.

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #7457: Archer Guard not gaining any Damage from Agility before 50
### Prerequisites 

- [no] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [yes] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [yes] I made sure that **this issue is not a duplicate of any existing issue.**

### Context 

- Minecraft Version: 1.16.5
- MineColonies Version: 0.14.310 Alpha universal
- Structurize Version: 0.13.202 Alpha universal
- Related Mods and their Versions: All The Mods 6 modpack

## Expected behavior

Archer Guard dealing at least 5 Attack Damage at 20 Agility (If he would deal 1 Damage [0.5 Hearts] at 0-4 Agility)

## Actual behavior

The Archer damage does not increase whatsoever in the levels at or below 20. It doesnt matter what bow he is using (Tested Vanilla bow and Dragon Bone Bow from ice and fire)

https://youtu.be/aVZNTPGiJM4 

## Steps to reproduce the issue

1.  Use a Guard
2.  Assign it as Archer
3.  Its damage doesnt Increase probably before reaching 50 Agility

### Logs

- latest.log: 
- crashlog: 

### Notes

---
## Viewers

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### PR DESCRIPTION
**Pull request by someaddons**
Closes #7450
Closes #6841
Closes #7457

# Changes proposed in this pull request:
Adds specific combat AI's for dealing with threattables, which now both guards and raiders use
Existing fighting AI's got refactored into those combat AI's
CombatAI's now are in a seperate AI which gets integrated
Raiders and guards are now better at target choosing, and both able to ignore unreachable targets, or those who took too long to reach. 

Review please

### CODE REVIEW

File: src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java [hunk 1]
```diff
@@ -479,7 +494,7 @@ public void die(@NotNull final DamageSource cause)
     @Override
     public boolean hurt(@NotNull final DamageSource damageSource, final float damage)
     {
-        if (damageSource.getEntity() instanceof LivingEntity)
+        if (damageSource.getEntity() instanceof LivingEntity && !(damageSource.getEntity() instanceof AbstractEntityMinecoloniesMob))
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java:510]: Oh, so no infighting anymore? =P

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/KnightCombatAI.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    /**
+     * Check if the guard can protect himself with a shield And if so, do it.
+     *
+     * @return The next IAIState.
+     */
+    protected IAIState attackProtect()
+    {
+        final int shieldSlot = InventoryUtils.findFirstSlotInItemHandlerWith(user.getInventoryCitizen(), Items.SHIELD);
+
+        if (target != null && target.isAlive() && nextAttackTime - user.level.getGameTime() >= 8)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/KnightCombatAI.java]: >= 8? Magi cnumber

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/KnightCombatAI.java]: you can merge those two ifs.

File: src/main/java/com/minecolonies/coremod/entity/ai/combat/AttackMoveAI.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+     */
+    private IState move()
+    {
+        if (!checkForTarget())
+        {
+            return CombatAIStates.NO_TARGET;
+        }
+
+        final ThreatTableEntry nextTarget = user.getThreatTable().getTarget();
+        if (nextTarget == null)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/combat/AttackMoveAI.java:62]: ?

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/pathjobs/PathJobRandomPos.java [hunk 4]
```diff
@@ -116,7 +114,8 @@ protected double computeHeuristic(@NotNull final BlockPos pos)
     @Override
     protected boolean isAtDestination(@NotNull final Node n)
     {
-        if ((minDistFromStart ==  1 || start.distSqr(n.pos) > minDistFromStart * minDistFromStart) && isWalkableSurface(world.getBlockState(n.pos.below()), n.pos.below()) == SurfaceType.WALKABLE
+        if (random.nextInt(10) == 0 && (start.distSqr(n.pos) > minDistFromStart * minDistFromStart)
+              && isWalkableSurface(world.getBlockState(n.pos.below()), n.pos.below()) == SurfaceType.WALKABLE
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/pathjobs/PathJobRandomPos.java:118]: random?

File: src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java (hunk 1 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/entity/mobs/AbstractEntityMinecoloniesMob.java:510]: ah yea xD

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/pathjobs/PathJobRandomPos.java (hunk 4 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/pathfinding/pathjobs/PathJobRandomPos.java:118]: ye to find a more spread out more random position

**Review by Raycoms — DISMISSED** [review]

**Pull request merged**